### DISCUSSION TITLE
Custom logo and SDCard flashing

### DISCUSSION THREAD

**Discussion by DarkCowMoo** [OE4T/meta-tegra#1026] [Q&A]
Hi,
First of all, thanks for all the effort in making this BSP layer. It looks like a lot of effort was sunk into this project which is highly appreciated.

I am having few difficulties which hopefully can be easily solved -

1. After following the image creating guide and creating a wic.gz file, I am trying to edit the /boot/extlinux/extlinux.conf file. 
$ gunzip -c core-image-base-jetson-xavier-nx-devkit.wic.gz | sudo dd of=/dev/sda1 bs=10M status=progress
Unfortunately, it does not exists. Basically, I cannot access the SDcard after the above line.
I can reformat it to access the SDCard regulardly so it is not ruined.

2. I tried to build a custom layer which edits the boot logo to try and understand better how Yocto works. This task has posed a bit of confusion and I am not entirely sure if I should add this layer to a new image, or to core-image-base which is what meta-tegra support. In addition, when adding to either images, I am facing an error -

IMAGE_INSTALL:append = " xtra-bootlogo"  ->

ERROR: Task do_populate_sdk in /home/shauli/Git/poky/meta-xtra/recipes-core/images/xtra-l4t.bb rdepends upon non-existent task do_package_write_deb in /home/shauli/Git/poky/meta-xtra/recipes-logo/xtra-bootlogo/xtra-bootlogo.bb

I basically copied the custom-logo.bb file. If I remove the '$ inherit nopackages', the project still fails with package not found.

E: Package 'xtra-bootlogo' has no installation candidate

Layer formation -

meta                  /home/shauli/Git/poky/meta                5
meta-poky             /home/shauli/Git/poky/meta-poky           5
meta-yocto-bsp        /home/shauli/Git/poky/meta-yocto-bsp      5
meta-tegra            /home/shauli/Git/poky/meta-tegra          5
meta-xtra             /home/shauli/Git/poky/meta-xtra           6

File formation - 

meta-xtra/
├── conf
│   └── layer.conf
├── COPYING.MIT
├── README
├── recipes-core
│   └── images
│       └── xtra-l4t.bb
└── recipes-logo
    ├── xtra-bootlogo
    │   ├── bootlogo.inc
    │   ├── files
    │   │   └── xtra_logo.tar.gz
    │   └── xtra-bootlogo.bb

Apologies for the rather basic questions. I am starting to get a grip on how things work with yocto and the meta bsp layers, but it is still rather difficult to proceed. 

Thanks,
Shauli

**Comment by dwalkes** [comment]
@DarkCowMoo
> After following the image creating guide and creating a wic.gz file

Can you link the image creating guide you are using?  Did you notice https://github.com/OE4T/meta-tegra/wiki/Flashing-the-Jetson-Dev-Kit ? 

WIC support was discussed a bit in https://github.com/OE4T/meta-tegra/discussions/620 but it's not currently something which is supported for the reasons discussed there.

> I tried to build a custom layer which edits the boot logo

Did you see the discussion at https://github.com/OE4T/meta-tegra/discussions/670?  This one gives an example of customizing the `bootlogo-custom`  provider to provide a custom boot logo.

**Reply by DarkCowMoo** [reply]
Hi Dan,
Thank you for the quick answer. 
I was following
https://developer.ridgerun.com/wiki/index.php/Yocto_Support_for_NVIDIA_Jetson_Platforms_-_Flashing_the_Jetson_Platform
I'll checkout your guide.

Regarding the custom logo -
I haven't followed thread #670. I was just trying to build a new layer over meta-tegra by simply copying the original bb layer and pointing to my own tar file. I imagined it would be simple enough. 

Thanks, 
Shauli

**Reply by dwalkes** [reply]
Thanks @DarkCowMoo I don't have experience with those instructions but I've pinged Ridgerun folks to see if they could weigh in.

> I was just trying to build a new layer over meta-tegra by simply copying the original bb layer and pointing to my own tar file

I haven't dug into the recipe enough to know exactly what you are missing, but in general you should avoid copying and append to modify recipes instead.  See https://docs.yoctoproject.org/dev-manual/common-tasks.html#following-best-practices-when-creating-layers

**Reply by DarkCowMoo** [reply]
Hi @dwalkes ,
Thanks for the answers.
I am following the flashing guide you pointed to. Jetson Xavier seems to be in a bootloop so I guess something went wrong. 

Regarding the recipe, I pasted it below. I wouldn't say it is a complete copy as I do need the 'cooking' part. #670 seems to be using SPLASH which I thought is a different method than what Nvidia does.  

DESCRIPTION = "Deploy a custom boot screen for L4T bootloader, overwriting Nvidia's logo"

LICENSE = "MIT"
LIC_FILES_CHKSUM = "file://${COMMON_LICENSE_DIR}/MIT;md5=0835ade698e0bcf8506ecda2f7b4f302"

COMPATIBLE_MACHINE = "(tegra)"

SOC_FAMILY ??= "tegra186"
DEPENDS = "${SOC_FAMILY}-flashtools-native lz4-native"

inherit deploy nopackages

PROVIDES += "virtual/xtra-bootlogo"

SRC_URI = "file://xtra_logo.tar.gz"
SRC_URI[sha256sum] = "3fa376841c81c69c5558714385195dbe492407a78d29c07e5ce89ad4b74e69b2"

S = "${WORKDIR}"
# One of 'raw' or 'cooked'
PACKAGECONFIG ?= "cooked"

PACKAGECONFIG[raw] = ",,imagemagick-native,,,cooked"
PACKAGECONFIG[cooked] = ",,,,,raw"

# Input input graphic to be scaled
RAW_GRAPHIC_FILE ?= "Skylord_Logo.png"

# Config file definting pregenerated bitmap files
COOKED_CONFIG_FILE ?= "config_file"

BMP_BLOB = "xtra_bootlogo-${MACHINE}-${PV}-${PR}.blob"
BMP_SYMLINK = "bootlogo-${MACHINE}.blob"

include bootlogo.inc

FLASHTOOLS_DIR = "${SOC_FAMILY}-flash"
FLASHTOOLS_DIR:tegra194 = "tegra186-flash"

do_compile() {

    LIST=""

    if [ "${@bb.utils.contains("PACKAGECONFIG", "raw", "1", "0", d)}" = "1" ]; then
        if [ ! -s "${S}/${RAW_GRAPHIC_FILE}" ]; then
            bbfatal "RAW_GRAPHIC_FILE must exist"
        fi

        cook_bitmaps "${S}/${RAW_GRAPHIC_FILE}"
    else
        if [ ! -s "${S}/${COOKED_CONFIG_FILE}" ]; then
            bbfatal "COOKED_CONFIG_FILE must exist"
        fi

        while read _line
        do
            _line="${S}/${_line}"
            LIST="${LIST}${_line};"
        done < ${S}/${COOKED_CONFIG_FILE}
    fi

    # NB that '%?' gets rid of the trailing ';' because BMP_generator blows up
    LIST="${LIST%?}"

    rm -f ${B}/bmp.blob
    rm -f ${B}/bmp-compressed.blob

    OUT=${B} ${STAGING_BINDIR_NATIVE}/${FLASHTOOLS_DIR}/BMP_generator_L4T.py -t bmp -e "${LIST}" -v 0

    if [ ! -s ${B}/bmp.blob ]; then
        bbfatal "BMP_generator_L4T.py failed to create bmp.blob"
    fi

    compress_blob "${B}/bmp.blob" "${B}/bmp-compressed.blob"

    if [ ! -s ${B}/bmp-compressed.blob ]; then
        bbfatal "failed to create compressed bmp-compressed.blob"
    fi
}

do_deploy() {
    install -d ${DEPLOYDIR}
    install -m 0644 ${B}/bmp-compressed.blob ${DEPLOYDIR}/${BMP_BLOB}
    # Should overwrite Nvidia's symlink
    ln -sf ${BMP_BLOB} ${DEPLOYDIR}/${BMP_SYMLINK}
}

PACKAGE_ARCH = "${MACHINE_ARCH}"
do_install[noexec] = "1"

addtask deploy before do_build after do_install

**Reply by dwalkes** [reply]
> I am following the flashing guide you pointed to. Jetson Xavier seems to be in a bootloop so I guess something went wrong.

If you haven't already, make sure you try using https://github.com/OE4T/tegra-demo-distro to build and flash an image successfully.  If you still have problems with the flash script please share the `MACHINE` and branch you are using.

**Comment by madisongh** [comment]
The custom bootlogo recipe doesn't install any files into the rootfs, so there are no packages generated for it.  Make sure your image recipe does not try to install that package.  Instead, set `PREFERRED_PROVIDER_virtual/bootlogo` to point to your custom bootlogo recipe.  Bear in mind, too, that we don't really support WIC images for the Jetson targets.  You should be using the `dosdcard.sh` script in the `tegraflash` package; see [this wiki page](https://github.com/OE4T/meta-tegra/wiki/Flashing-the-Jetson-Dev-Kit) for instructions.  Your custom bootlogo should be in that package as well if you've set the above-mentioned variable.

Also note that the bootlogo is stored in the SPI flash on the Xavier NX modules, so you'll need to go through the steps mentioned on the wiki page to re-flash the boot partitions there in order to see your custom logo.

**Reply by DarkCowMoo** [reply]
Thanks for the answer. I am facing an error but I think this one is just related to recovery mode. Will update if I face issues.
./doflash.sh --spi-only
ERR: could not retrieve chip ID

Regarding the recipe, for some reason I thought the original meta-tegra installs the logo package which is cleary wrong and probably the reason I faced this error. I will try to add PREFERRED_PROVIDER_virtual/bootlogo.
There is some much to learn about Yocto, regardless of the meta-tegra, which I find both confusing and frustrating. 

Thanks for the answer!
